### PR TITLE
fix(aliasing): added new resolvedName property to typeIdentifier object for import aliased declrations

### PR DESCRIPTION
**Pull request by sanketshevkar**
# PartialyCloses https://github.com/accordproject/concerto/issues/972

Added to new optional`resolvedName` property to `TypeIdenifier`. This is to preserve the aliased `name` property which will have the local name for the aliased type. `resolvedName` will contain the origonal type name. It will get added only for the declrations ot types that "imported **and** aliased".

With the current implementation the value of the `name` property is different for a normal ast (local name) vs a resolved ast (original type name). This can be very confusing.

The part that would close this issue would be adding unit tests to cover the edge cases in concerto packages.

### Related Issues
- Issue #32 
- Pull Request #<NUMBER>

### Author Checklist
- [ ] Ensure you provide a [DCO sign-off](https://github.com/probot/dco#how-it-works) for your commits using the `--signoff` option of git commit.
- [ ] Vital features and changes captured in unit and/or integration tests
- [ ] Commits messages follow [AP format](https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format)
- [ ] Extend the documentation, if necessary
- [ ] Merging to `main` from `fork:branchname`

### CODE REVIEW

File: lib/metamodelutil.js [hunk 1]
```diff
@@ -89,10 +89,11 @@ function createNameTable(priorModels, metaModel) {
                 if (!findDeclaration(modelFile, type)) {
                     throw new Error(`Declaration ${type} in namespace ${namespace} not found`);
                 }
-                table[localName] = {namespace, name: type};
+                table[localName] = localName !== type ? {namespace, name: localName, resolvedName: type} : {namespace, name: type};
             });
         } else {
             (modelFile.declarations || []).forEach((decl) => {
+                // What is this used for?
```
Comment (ekarademir) [inline lib/metamodelutil.js]: Why add a comment here? This comment doesn't explain the code or notify anything. Maybe just add a comment to the PR? Or raise it in Discord?

Could you remove it before you merge this?

**Review by jonathan-casey — APPROVED** [review]

File: lib/metamodelutil.js (hunk 1 above)
Comment (ekarademir) [reply lib/metamodelutil.js]: The code is adding types defined in the current CTO to the types table.

File: lib/metamodelutil.js [hunk 2]
```diff
@@ -147,6 +148,9 @@ function resolveTypeNames(metaModel, table) {
             const name = metaModel.superType.name;
             metaModel.superType.namespace = resolveName(name, table);
             metaModel.superType.name = table[name].name;
+            if (table[name]?.resolvedName) {
```
Comment (ekarademir) [inline lib/metamodelutil.js:150]: Wouldn't you want `superType.name` to be the `resolvedName`? `name` for an aliased import is still local. So if some code is using `${superType.namespace}.${superType.name}`, then that wouldn't resolve to anything, unless it's the resolved name.

FQNs should be global, no?

**Review by ekarademir — CHANGES_REQUESTED** [review]

File: lib/metamodelutil.js (hunk 1 above)
Comment (sanketshevkar) [reply lib/metamodelutil.js]: Sorry I had forgot to remove the comment from the older PR.

File: lib/metamodelutil.js (hunk 2 above)
Comment (sanketshevkar) [reply lib/metamodelutil.js:150]: That's the tricky bit there. We wanted to avoid the name property to change its value when we resolve it. That would need a lot of changes in concerto-core/cto/vocab because these libraries can't really figure out if a model passed to them is resolved or not.

If we set `superType.name` to `resolvedName` then we would need to add check everywhere else to see if the model present in manager is resolved or not.

With this new property resolvedName we don't have to make any changes to concerto packages, I've manually tested and have being adding unit tests and they seem to work fine as they don't really use the namespace and resolvedName property.

For other downstream systems we've decided to rely communications and document the expection neatly.
_If they you are dealing with an resolved ast of a model and you encounter the resolvedName property then use that to refer the type from the namespace value of the namespace property._

File: lib/metamodelutil.js [hunk 3]
```diff
@@ -89,7 +89,7 @@ function createNameTable(priorModels, metaModel) {
                 if (!findDeclaration(modelFile, type)) {
                     throw new Error(`Declaration ${type} in namespace ${namespace} not found`);
                 }
-                table[localName] = {namespace, name: type};
+                table[localName] = localName !== type ? {namespace, name: localName, resolvedName: type} : {namespace, name: type};
```
Comment (ekarademir) [inline lib/metamodelutil.js:92]: So here, if the imported type is aliased, you are replacing the name of the type with its local name.

Which will then be used to assign the super type of the concept here https://github.com/accordproject/concerto/blob/85366be18b5deb3efaa6acfbc7319b14a0219c05/packages/concerto-core/lib/introspect/classdeclaration.js#L70

Which would then fail to resolve, because the runtime will try to resolve the type using that name here https://github.com/accordproject/concerto/blob/85366be18b5deb3efaa6acfbc7319b14a0219c05/packages/concerto-core/lib/introspect/classdeclaration.js#L163

because it won't exist.

**Review by ekarademir — APPROVED** [review]
